tests/test_mtype_tokenizer.py

```diff
@@ -0,0 +1,188 @@
+import warnings
+
+import pytest
+
+from amads.algorithms.mtype_tokenizer import FantasticTokenizer, MType
+from amads.algorithms.ngrams import NGramCounter
+from amads.core.basics import Note, Score
+from amads.melody.segment import fantastic_segmenter
+
+
+def test_mtype_tokenizer():
+    melody_1 = Score.from_melody(
+        pitches=[62, 64, 65, 67, 64, 60, 62],  # D4, E4, F4, G4, E4, C4, D4
+        durations=[
+            0.5,
+            0.5,
+            0.5,
+            0.5,
+            1.0,
+            0.5,
+            0.5,
+        ],  # quarter, eighth, quarter, eighth, eighth notes
+    )
+
+    # Initialize tokenizer and counter
+    tokenizer = FantasticTokenizer()
+    ngrams = NGramCounter()
+
+    # Tokenize melody and count n-grams
+    tokens = tokenizer.tokenize(melody_1)
+
+    # Count bigrams
+    ngrams.count_ngrams(tokens, n=2)
+    for ngram in ngrams.get_counts():
+        assert len(ngram) == 2
+
+    # Reset counter, count trigrams
+    ngrams.reset()
+    ngrams.count_ngrams(tokens, n=3)
+    for ngram in ngrams.get_counts():
+        assert len(ngram) == 3
+
+    # Test that n-grams cannot be counted if method is larger than sequence length
+    with pytest.raises(ValueError):
+        ngrams.count_ngrams(tokens, n=10)
+
+    # Test that n-grams cannot be counted if method n = 0
+    with pytest.raises(ValueError):
+        ngrams.count_ngrams(tokens, n=0)
+
+    # Test that n=None returns all n-grams
+    ngrams.reset()
+    ngrams.count_ngrams(tokens, n=None)
+    all_counts = ngrams.get_counts()
+    # Test each n-gram length by comparing to individual n-gram counts
+    max_length = max(len(ngram) for ngram in all_counts)
+    for n in range(1, max_length + 1):
+        ngrams.reset()
+        ngrams.count_ngrams(tokens, n=n)
+        n_gram_counts = ngrams.get_counts()
+        # Check that all n-grams of length n in n_gram_counts are also in all_counts
+        for ngram in n_gram_counts:
+            assert ngram in all_counts
+            assert n_gram_counts[ngram] == all_counts[ngram]
+
+    # Test with a second melody using a different phrase gap
+    tokenizer = FantasticTokenizer()
+    ngrams = NGramCounter()
+
+    # Create a test melody which will be segmented into 2 phrases
+    melody_2 = Score.from_melody(
+        # Twinkle Twinkle Little Star
+        pitches=[60, 60, 67, 67, 69, 69, 67, 65, 65, 64, 64, 62, 62, 60],
+        durations=[
+            0.5,
+            0.5,
+            0.5,
+            0.5,
+            0.5,
+            0.5,
+            1.0,
+            0.5,
+            0.5,
+            0.5,
+            0.5,
+            0.5,
+            0.5,
+            0.5,
+        ],
+    )
+
+    # Test that the melody is segmented into 2 phrases
+    segments = fantastic_segmenter(melody_2, phrase_gap=0.75, units="quarters")
+    # Extract notes from each segment for counting
+    segment_notes = [list(segment.find_all(Note)) for segment in segments]
```

Comment:
This has the same problem we talked about before, where ngrams are going to formed across segment boundaries.